File: src/deadlock_build_sync/offline/discovery_snapshot.py
"""Save the complete discovery family before validation and verify it for reuse."""

import json
from hashlib import file_digest
from typing import cast

from deadlock_build_sync.artifacts import atomic_write_bytes, atomic_write_json
from deadlock_build_sync.build_evidence import METHOD_VERSION
from deadlock_build_sync.snapshot import sha256_json
from deadlock_build_sync.value_validation import (
    integer,
    object_dict,
    require_object_dict,
)

from .config import RunPaths
from .discovery_guide_groups import assign_guide_group_ids
from .discovery_types import FrozenHeroDiscovery


def calculate_discovery_source_identity(paths: RunPaths) -> dict[str, object]:
    hashes = {}
    for name in (
        "manifest.json",
        "raw/analysis.duckdb",
        "raw/heroes.json",
        "raw/items.json",
        "raw/items-all.json",
        "raw/ranks.json",
        "raw/patches.json",
    ):
        with (paths.run / name).open("rb") as source:
            hashes[name] = file_digest(source, "sha256").hexdigest()
    return {"schema_version": 1, "method_version": METHOD_VERSION, "files": hashes}


def require_discovery_source_identity(
    paths: RunPaths, expected: dict[str, object]
) -> None:
    if calculate_discovery_source_identity(paths) != expected:
        raise ValueError("Discovery source identity differs. Start a new --run-id.")


def save_discovery_snapshot(
    paths: RunPaths,
    frozen: dict[int, FrozenHeroDiscovery],
    groups: dict[int, dict[str, str]],
    source_identity: dict[str, object],
) -> None:
    require_discovery_source_identity(paths, source_identity)
    digest = sha256_json(frozen)
    atomic_write_bytes(
        paths.run / f"discovery-nominations-{digest[:16]}.json",
        json.dumps(frozen, allow_nan=False).encode(),
    )
    atomic_write_json(
        paths.run / f"guide-groups-{digest[:16]}.json",
        {
            "frozen_sha256": digest,
            "groups": groups,
            "source_identity": source_identity,
        },
    )


def load_discovery_snapshot(
    paths: RunPaths, heroes: list[dict[str, object]]
) -> tuple[dict[int, FrozenHeroDiscovery], dict[int, dict[str, str]]]:
    candidates = list(paths.run.glob("discovery-nominations-*.json"))
    if len(candidates) != 1:
        raise ValueError("Resume requires one complete discovery snapshot")
    document = require_object_dict(
        json.loads(candidates[0].read_text(encoding="utf-8"))
    )
    frozen = {
        integer(hero): cast("FrozenHeroDiscovery", require_object_dict(record))
        for hero, record in document.items()
    }
    if set(frozen) != {integer(hero["id"]) for hero in heroes}:
        raise ValueError("Discovery snapshot does not cover the requested heroes")
    digest = sha256_json(frozen)
    if candidates[0].name != f"discovery-nominations-{digest[:16]}.json":
        raise ValueError("Discovery snapshot fingerprint does not match")
    groups_path = paths.run / f"guide-groups-{digest[:16]}.json"
    group_document = require_object_dict(
        json.loads(groups_path.read_text(encoding="utf-8"))
    )
    source_identity = object_dict(group_document.get("source_identity"))
    if source_identity is None:
        raise ValueError("Discovery has no source identity. Start a new --run-id.")
    require_discovery_source_identity(paths, source_identity)
    groups = {
        hero: assign_guide_group_ids(report["rows"]) for hero, report in frozen.items()
    }
    if group_document != {
        "frozen_sha256": digest,
        "groups": {str(hero): group for hero, group in groups.items()},
        "source_identity": source_identity,
    }:
        raise ValueError("Discovery snapshot guide groups do not match")
    return frozen, groups

File: crates/deadlock-analysis/src/discovery_snapshot.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::{
    Error, Result, atomic_write_json, canonical_json, file_sha256, fingerprint, integer, read_json,
    sha256,
};
use deadlock_guides::CURRENT_METHOD_VERSION;
use serde_json::{Value, json};

use crate::config::RunPaths;
use crate::core_discovery::assign_group_ids;
use crate::discovery_models::FrozenHero;

pub type FrozenRoster = BTreeMap<u64, FrozenHero>;
pub type GuideGroups = BTreeMap<u64, BTreeMap<String, String>>;

pub fn source_identity(paths: &RunPaths) -> Result<Value> {
    let mut hashes = BTreeMap::new();
    for name in [
        "manifest.json",
        "raw/analysis.duckdb",
        "raw/heroes.json",
        "raw/items.json",
        "raw/items-all.json",
        "raw/ranks.json",
        "raw/patches.json",
    ] {
        hashes.insert(name, file_sha256(&paths.run.join(name))?);
    }
    Ok(json!({"schema_version":1,"method_version":CURRENT_METHOD_VERSION,"files":hashes}))
}

pub fn require_source_identity(paths: &RunPaths, expected: &Value) -> Result<()> {
    if source_identity(paths)? != *expected {
        return Err(Error::new(
            "Discovery source identity differs. Start a new --run-id.",
        ));
    }
    Ok(())
}

pub fn roster_fingerprint(frozen: &FrozenRoster) -> Result<String> {
    let mut bytes = vec![b'{'];
    for (index, (hero, record)) in frozen.iter().enumerate() {
        if index > 0 {
            bytes.push(b',');
        }
        bytes.extend(serde_json::to_vec(&hero.to_string())?);
        bytes.push(b':');
        bytes.extend(canonical_json(&serde_json::to_value(record)?)?);
    }
    bytes.push(b'}');
    Ok(sha256(&bytes))
}

pub fn save_snapshot(
    paths: &RunPaths,
    frozen: &FrozenRoster,
    groups: &GuideGroups,
    source: &Value,
) -> Result<String> {
    require_source_identity(paths, source)?;
    let digest = roster_fingerprint(frozen)?;
    atomic_write_json(
        &paths
            .run
            .join(format!("discovery-nominations-{}.json", &digest[..16])),
        &serde_json::to_value(frozen)?,
    )?;
    atomic_write_json(
        &paths
            .run
            .join(format!("guide-groups-{}.json", &digest[..16])),
        &json!({"frozen_sha256":digest,"groups":groups,"source_identity":source}),
    )?;
    Ok(digest)
}

pub fn load_snapshot(paths: &RunPaths, heroes: &[Value]) -> Result<(FrozenRoster, GuideGroups)> {
    let mut candidates = Vec::new();
    for entry in std::fs::read_dir(&paths.run)? {
        let entry = entry?;
        let name = entry.file_name();
        let name = name.to_string_lossy();
        if name.starts_with("discovery-nominations-") && name.ends_with(".json") {
            candidates.push(entry.path());
        }
    }
    if candidates.len() != 1 {
        return Err(Error::new(
            "Resume requires one complete discovery snapshot",
        ));
    }
    let document = read_json(&candidates[0])?;
    let frozen: FrozenRoster = serde_json::from_value(document)?;
    let expected = heroes
        .iter()
        .map(|hero| integer(hero, "id"))
        .collect::<Result<BTreeSet<_>>>()?;
    if frozen.keys().copied().collect::<BTreeSet<_>>() != expected {
        return Err(Error::new(
            "Discovery snapshot does not cover the requested heroes",
        ));
    }
    let digest = roster_fingerprint(&frozen)?;
    if candidates[0].file_name().and_then(|name| name.to_str())
        != Some(&format!("discovery-nominations-{}.json", &digest[..16]))
    {
        return Err(Error::new("Discovery snapshot fingerprint does not match"));
    }
    let group_document = read_json(
        &paths
            .run
            .join(format!("guide-groups-{}.json", &digest[..16])),
    )?;
    require_source_identity(paths, &group_document["source_identity"])?;
    let groups = frozen
        .iter()
        .map(|(hero, report)| Ok((*hero, assign_group_ids(&report.rows)?)))
        .collect::<Result<GuideGroups>>()?;
    let expected = json!({"frozen_sha256":digest,"groups":groups,"source_identity":group_document["source_identity"]});
    if fingerprint(&group_document)? != fingerprint(&expected)? {
        return Err(Error::new("Discovery snapshot guide groups do not match"));
    }
    Ok((frozen, groups))
}
